### PR TITLE
Add label_smoothing function into CrossEntropyLoss

### PR DESCRIPTION
**Pull request by MqLeet**
### PR types
New features

### PR changes
APIs

### Description
1. add label_smoothing parameter into paddle.nn.CrossEntropyLoss
2. add label_smoothing function into paddle.nn.functional.crossentropyloss. Now this function can accepy both one-hot label and Integer-form label when `soft_label` is True
3. add 8 test cases for label smoothing function in test_cross_entropy_loss.py. All cases are OK.

### CODE REVIEW

File: python/paddle/nn/functional/loss.py [hunk 1]
```diff
@@ -2453,6 +2453,7 @@ def cross_entropy(
     axis=-1,
     use_softmax=True,
     name=None,
+    label_smoothing=0.0,
```
Comment (zhwesky2010) [inline python/paddle/nn/functional/loss.py:2506]: 这个放在name前面吧，name相当于未使用的，不会有不兼容问题

File: python/paddle/nn/functional/loss.py [hunk 2]
```diff
@@ -2597,6 +2602,11 @@ def cross_entropy(
             If :attr:`reduction` is ``'none'``, the unreduced loss is returned.
             Default is ``'mean'``.
         soft_label (bool, optional): Indicate whether label is soft. Default is ``False``.
+        label_smoothing (float, optional): A float in [0.0, 1.0], only valid when soft_label=True.
```
Comment (zhwesky2010) [inline python/paddle/nn/functional/loss.py]: 竞品是没有soft_label功能，但也有label_smoothing，所以这个只能soft_label才能生效吗

File: python/paddle/nn/functional/loss.py [hunk 3]
```diff
@@ -2641,7 +2651,7 @@ def cross_entropy(
             # Tensor(shape=[], dtype=float64, place=Place(gpu:0), stop_gradient=True,
             #        5.34043430)
 
-        .. code-block:: python
+        .. code-block:: python without label_smoothing
```
Comment (zhwesky2010) [inline python/paddle/nn/functional/loss.py]: 这个格式可能不对，这个python注释写到下面，但不要放到.. code-block:: python里：

比如：
\# case1：without label_smoothing
\# case2: with label_smoothing

File: python/paddle/nn/layer/loss.py [hunk 4]
```diff
@@ -345,6 +354,33 @@ class CrossEntropyLoss(Layer):
             Tensor(shape=[], dtype=float64, place=Place(cpu), stop_gradient=True,
             1.14554912)
 
+        .. code-block:: python
+
+            >>> # soft labels with label_smoothing
```
Comment (zhwesky2010) [inline python/paddle/nn/layer/loss.py]: 这种写法可能不符合目前的文档规范了

File: python/paddle/nn/layer/loss.py [hunk 5]
```diff
@@ -322,7 +331,7 @@ class CrossEntropyLoss(Layer):
 
         .. code-block:: python
 
-            >>> # soft labels
+            >>> # soft labels without label_smoothing
```
Comment (zhwesky2010) [inline python/paddle/nn/layer/loss.py]: 同上

File: python/paddle/nn/layer/loss.py [hunk 6]
```diff
@@ -356,6 +392,7 @@ def __init__(
         axis=-1,
         use_softmax=True,
         name=None,
+        label_smoothing=0.0,
```
Comment (zhwesky2010) [inline python/paddle/nn/layer/loss.py:386]: 放到name前面

File: python/paddle/nn/layer/loss.py [hunk 7]
```diff
@@ -365,6 +402,7 @@ def __init__(
         self.axis = axis
         self.use_softmax = use_softmax
         self.name = name
+        self.label_smoothing = label_smoothing
```
Comment (jeff41404) [inline python/paddle/nn/layer/loss.py:396]: 这行也移动到name前面

File: python/paddle/nn/functional/loss.py (hunk 2 above)
Comment (MqLeet) [reply python/paddle/nn/functional/loss.py]: 是的。 在这个Pr下面，只有soft_label=True时，对label_smoothing这个参数所做的修改才会生效。

理由：
因为在当前的CrossEntropyLoss函数中包含有`soft_label`这个参数,默认为False
如果soft_label=False，则接受integer形式的label。
如果soft_label=True，则需要用户自行提供smoothing之后的label，形状需要和input一致。

在当前这个Pr中：
如果设置soft_label=False, 则仍然和之前的API兼容，不做变动
如果设置soft_label=True，则此时可以接受两种形式的输入：one-hot label 和 integer label, 在函数内会自动把label都转变成和input同样的形状。
  - 如果设置label_smoothing=0.0, 则函数不会对用户的输入做label_smoothing. 此时期望用户的输入就是soft label
  - 如果设置label_smoothing>0.0, 则函数会对label做一次label_smoothing. 此时期望用户的输入是 hard label

File: python/paddle/nn/layer/loss.py (hunk 4 above)
Comment (MqLeet) [reply python/paddle/nn/layer/loss.py]: 在nn.layer.loss和nn.functional.loss中我看到了两种不同形式的 注释中的example， 是应该按照nn.functional.loss的写法来写吗？

Comment (zhwesky2010) [reply python/paddle/nn/layer/loss.py]: 按最新的写，看一下主流是哪些，少数可能是之前未更新

File: python/paddle/nn/functional/loss.py (hunk 3 above)
Comment (MqLeet) [reply python/paddle/nn/functional/loss.py]: 已修改

File: python/paddle/nn/layer/loss.py (hunk 4 above)
Comment (MqLeet) [reply python/paddle/nn/layer/loss.py]: 已修改

File: python/paddle/nn/layer/loss.py (hunk 7 above)
Comment (MqLeet) [reply python/paddle/nn/layer/loss.py:396]: 已修改

File: python/paddle/nn/layer/loss.py (hunk 6 above)
Comment (MqLeet) [reply python/paddle/nn/layer/loss.py:386]: 已修改

File: python/paddle/nn/functional/loss.py [hunk 8]
```diff
@@ -2664,8 +2674,35 @@ def cross_entropy(
                                                                     weight=weight,
                                                                     reduction=reduction)
             print(paddle_loss_mean)
-            # Tensor(shape=[], dtype=float64, place=Place(gpu:0), stop_gradient=True,
-            #        1.11043464)
+            # Tensor(shape=[], dtype=float64, place=Place(cpu), stop_gradient=True,
+            #        1.12801195)
+
+        .. code-block:: python
```
Comment (zhwesky2010) [inline python/paddle/nn/functional/loss.py]: 这个应该也不用重新起一个代码块，直接放在一个代码块里，case1、case2就可以

File: python/paddle/nn/functional/loss.py [hunk 9]
```diff
@@ -2580,8 +2581,12 @@ def cross_entropy(
             :math:`[N_1, N_2, ..., N_k]` or :math:`[N_1, N_2, ..., N_k, 1]`, k >= 1.
             the data type is int32, int64, float32, float64, where each value is [0, C-1].
 
-            2. If soft_label=True, the shape and data type should be same with ``input`` ,
-            and the sum of the labels for each sample should be 1.
+            2. If soft_label=True and label_smoothing = 0.0, the shape and data type
```
Comment (zhwesky2010) [inline python/paddle/nn/functional/loss.py]: no label_smoothing

File: python/paddle/nn/functional/loss.py [hunk 10]
```diff
@@ -2580,8 +2581,12 @@ def cross_entropy(
...
+            should be same with ``input`` , and the sum of the labels for each sample should be 1.
+
+            3. If label_smoothing > 0.0, soft_label will be ``True`` automatically, the shape and data type of ``label``
```
Comment (zhwesky2010) [inline python/paddle/nn/functional/loss.py]: If has `label_smoothing` , no matter what `soft_label` is, ... 后面这段描述感觉要写简单点，直接写input可以是什么形状和dtype，目前我感觉看不大清晰

File: python/paddle/nn/layer/loss.py [hunk 11]
```diff
@@ -304,48 +313,69 @@ class CrossEntropyLoss(Layer):
...
+            labels = paddle.uniform(shape, dtype='float64', min=0.1, max=1.0)
+            labels /= paddle.sum(labels, axis=axis, keepdim=True)
+            cross_entropy_loss = paddle.nn.loss.CrossEntropyLoss(
+                weight=weight, reduction=reduction, soft_label=True, label_smoothing=0.0)
+            dy_ret = cross_entropy_loss(logits, labels)
+            print(dy_ret)
+            # Tensor(shape=[], dtype=float64, place=Place(cpu), stop_gradient=True,
+            # 1.14554912)
+
+        .. code-block:: python
```
Comment (zhwesky2010) [inline python/paddle/nn/layer/loss.py]: 代码块可以合并为一个

File: python/paddle/nn/layer/loss.py [hunk 12]
```diff
@@ -304,48 +313,69 @@ class CrossEntropyLoss(Layer):
...
+            weight = paddle.rand([C], dtype='float64')
+
+            cross_entropy_loss = paddle.nn.loss.CrossEntropyLoss(
+                weight=weight, reduction=reduction)
+            dy_ret = cross_entropy_loss(input, label)
+            print(dy_ret)
+            # Tensor(shape=[], dtype=float64, place=Place(cpu), stop_gradient=True,
+            # 5.33697682)
 
         .. code-block:: python
```
Comment (zhwesky2010) [inline python/paddle/nn/layer/loss.py:332]: 代码块可以合并为一个

File: python/paddle/nn/functional/loss.py [hunk 13]
```diff
@@ -2698,6 +2735,21 @@ def cross_entropy(
             )
         )
 
+    if label_smoothing > 0.0:
+        soft_label = True
+        # converting the label to one-hot encoding
+        # for 1d case, converting label's shape from [N] to [N, C]
+        # for 2d case, converting label's shape from [N, d_1, ..., d_k] to [N, d_1, ..., d_k, C]
+        if input_dims != label_dims and input_dims - 1 == label_dims:
```
Comment (zhwesky2010) [inline python/paddle/nn/functional/loss.py]: input_dims != label_dims 这个条件还需要吗，与后面条件冗余吧

File: python/paddle/nn/functional/loss.py (hunk 10 above)
Comment (MqLeet) [reply python/paddle/nn/functional/loss.py]: image

请问这样的描述是否可行

File: python/paddle/nn/functional/loss.py (hunk 9 above)
Comment (MqLeet) [reply python/paddle/nn/functional/loss.py:2584]: 现在label的描述为：
        label (Tensor):
            1. If soft_label=False, the shape is
            :math:`[N_1, N_2, ..., N_k]` or :math:`[N_1, N_2, ..., N_k, 1]`, k >= 1.
            the data type is int32, int64, float32, float64, where each value is [0, C-1].

            2. If soft_label=True and no label_smoothing, the shape and data type
            should be same with ``input`` , and the sum of the labels for each sample should be 1.

            3. If has label_smoothing, (i.e. label_smoothing > 0.0), no matter what ``soft_label`` is, 
            the shape and data type of ``label`` could be either the situation 1 or situation 2.
            In other words, if label_smoothing > 0.0, the format of label could be one-hot label or integer label.
是否可行呢

File: python/paddle/nn/layer/loss.py (hunk 12 above)
Comment (MqLeet) [reply python/paddle/nn/layer/loss.py:332]: 您的意思是把hard labels和soft labels都放在同一个代码块中吗？

File: python/paddle/nn/functional/loss.py (hunk 8 above)
Comment (MqLeet) [reply python/paddle/nn/functional/loss.py]: image
这种注释写法是否可行呢

File: python/paddle/nn/functional/loss.py (hunk 13 above)
Comment (MqLeet) [reply python/paddle/nn/functional/loss.py]: 已修改

File: python/paddle/nn/functional/loss.py (hunk 9 above)
Comment (MqLeet) [reply python/paddle/nn/functional/loss.py]: image

File: python/paddle/nn/functional/loss.py (hunk 8 above)
Comment (zhwesky2010) [reply python/paddle/nn/functional/loss.py]: 是的，就是这样写

File: python/paddle/nn/functional/loss.py (hunk 10 above)
Comment (zhwesky2010) [reply python/paddle/nn/functional/loss.py]: 是的，就这么写

File: python/paddle/nn/functional/loss.py (hunk 8 above)
Comment (zhwesky2010) [inline python/paddle/nn/functional/loss.py:2680]: 这个你改了吗，为何要重新起一个 .. code-block:: python 代码块呢，所有示例代码放到一个里面就可以了

File: test/legacy_test/test_cross_entropy_loss.py [hunk 14]
```diff
@@ -810,6 +839,835 @@ def test_cross_entropy_loss_soft_2d_weight_mean(self):
...
+                    'input': self.logits,
+                    'label': self.labels,
+                    "weight": self.weight,
+                },
+                fetch_list=[ret],
+            )
+            self.assertIsNotNone(static_ret)
+        paddle.disable_static()
+
+        np.testing.assert_allclose(static_ret[0], expected, rtol=1e-05)
```
Comment (zhwesky2010) [inline test/legacy_test/test_cross_entropy_loss.py:1035]: 1. 这8个测试，能否考虑使用循环做一些代码复用，目前看代码的重合性较大
2. 反向似乎没有进行测试，一般一个OP的测试，前反向都是需要的，这个api之前也没有测反向吗

Comment (MqLeet) [reply test/legacy_test/test_cross_entropy_loss.py:1035]: 1. 这8个测试均是仿照该api之前的测试case来写的，之前的测试代码重合性较大
2. 这个api之前的测试case就没有测反向

Comment (zhwesky2010) [reply test/legacy_test/test_cross_entropy_loss.py:1035]: > 1. 这8个测试均是仿照该api之前的测试case来写的，之前的测试代码重合性较大
> 2. 这个api之前的测试case就没有测反向

这个相当于组合python api实现，实在没测反向也行吧

**Comment by zhwesky2010** [comment]
@MqLeet 这个PR的状态好像不对，不应该是Draft

**Review by zhwesky2010 — DISMISSED** [review]
LGTM

**Comment by MqLeet** [comment]
> 

Done

**Comment by MqLeet** [comment]
> @MqLeet 这个PR的状态好像不对，不应该是Draft

Done

**Review by jeff41404 — APPROVED** [review]
LGTM

**Review by tianshuo78520a — APPROVED** [review]

File: test/legacy_test/test_cross_entropy_loss.py [hunk 15]
```diff
@@ -23,6 +23,35 @@
 from paddle.base import Program, program_guard
 
 
+def label_smooth(label, C, epsilon, is_onehot=True):
+    """
+    Smooths the labels, commonly used in machine learning and deep learning to address label noise issues.
+
+    Args:
```
Comment (HackathonBot) [inline test/legacy_test/test_cross_entropy_loss.py:30]: 参考其他api添加参数的数据类型和是否optional吧

**Review by HackathonBot — COMMENTED** [review]
需要在 https://github.com/PaddlePaddle/docs/tree/develop/docs/api 对应目录修改中文api文档

File: test/legacy_test/test_cross_entropy_loss.py (hunk 15 above)
Comment (Ligoml) [inline test/legacy_test/test_cross_entropy_loss.py:30]: 参考其他api文档格式，需要在参数后注明支持的数据类型及是否可选（optional）

File: test/legacy_test/test_cross_entropy_loss.py [hunk 16]
```diff
@@ -23,6 +23,35 @@
...
+        label: Labels, of type np.ndarray, with shape (batch_size, C)
+        C: Number of classes
+        epsilon: Smoothing factor, should be greater than or equal to 0 and less than or equal to 1
+        is_onehot: Whether the labels are in one-hot encoding, default is True
+
+    Returns:
+        smooth_labels: Smoothed labels, of type np.ndarray, with the same shape as label
```
Comment (Ligoml) [inline test/legacy_test/test_cross_entropy_loss.py:37]: 这里不要加 `:` ，会有奇怪的官网解析，建议改成 `,`

File: test/legacy_test/test_cross_entropy_loss.py (hunk 15 above)
Comment (MqLeet) [reply test/legacy_test/test_cross_entropy_loss.py:30]: 这个不是api，是我在测试文件中写的用numpy计算label_smoothing的功能函数，并不对外开放

File: test/legacy_test/test_cross_entropy_loss.py (hunk 16 above)
Comment (MqLeet) [reply test/legacy_test/test_cross_entropy_loss.py:37]: 5555这个不是api，只是我在测试文件里写的功能函数，不对外开放

**Comment by MqLeet** [comment]
> 需要在 https://github.com/PaddlePaddle/docs/tree/develop/docs/api 对应目录修改中文api文档

Done

https://github.com/PaddlePaddle/docs/pull/6178

File: python/paddle/nn/functional/loss.py [hunk 17]
```diff
@@ -2673,7 +2683,7 @@ def cross_entropy(
     Examples:
         .. code-block:: python
 
-            >>> # hard labels
+            # hard labels
```
Comment (sunzhongkai588) [inline python/paddle/nn/functional/loss.py:2686]: 这边注释前面也得加 `>>>`

File: python/paddle/nn/layer/loss.py (hunk 12 above)
Comment (sunzhongkai588) [reply python/paddle/nn/layer/loss.py:332]: `hard` 和 `soft label` 的合并的话，看看可不可以把一些重复部分（代码、参数）给合并了，单纯的合到一起太长了，用户复制一大段

**Review by sunzhongkai588 — APPROVED** [review]
LGTM，后续文档部分单独提pr修改

**Review by nemonameless — APPROVED** [review]

**Review by zhwesky2010 — APPROVED** [review]

**Review by zhwesky2010 — APPROVED** [review]
